core/src/util/usFrameworkFactory.cpp

```diff
@@ -27,10 +27,46 @@
 
 namespace us {
 
+// A helper class to shut down the framework when all external
+// references (Bundle, BundleContext, etc.) are destroyed. It
+// also manages the lifetime of the CoreBundleContext instance.
+struct CoreBundleContextHolder
+{
+  CoreBundleContextHolder(std::unique_ptr<CoreBundleContext> ctx)
+    : ctx(std::move(ctx))
+  {}
+
+  ~CoreBundleContextHolder()
+  {
+    auto const state = ctx->systemBundle->state.load();
+    if (((Bundle::STATE_STARTING | Bundle::STATE_ACTIVE) & state) == 0)
+    {
+      // Call WaitForStop in case some did call Framework::Stop()
+      // but didn't wait for it. This joins with a potentially
+      // running framework shut down thread.
+      ctx->systemBundle->WaitForStop(std::chrono::milliseconds::zero());
+      return;
+    }
+
+    // Create a new CoreBundleContext holder, in case some event listener
+    // gets the system bundle and starts it again.
+    auto fwCtx = ctx.get();
+    std::shared_ptr<CoreBundleContext> holder(std::make_shared<CoreBundleContextHolder>(std::move(ctx)), fwCtx);
+    holder->SetThis(holder);
+    holder->systemBundle->Shutdown(false);
```

Comment:
I was aware of the limitations while executing code in `DllMain`. That was one reason why I was happy to move away from registering modules and calling their activators during static initialization of shared libraries.

You have a valid point here and I think a synchronous shut down makes sense in this case (we are waiting for it to complete anyways). But we still cannot ensure that client code that is executed in `Activator::Stop()` functions (executed during framework stop) plays by the rules for `DllMain`.

So
 1. I agree on "synchronous shutdown" during implicit framework destruction
 1. Question: Should we ensure that all CppMicroServices bundles can be stopped safely in the context of 1. ?
 1. This looks like a good topic for the "Best Practices" document